src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -34,7 +36,54 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
         return new ValueTask<int>(read, 0);
     }
 
-    public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    public async ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    {
+        if (data.IsSingleSegment)
+        {
+            return await WriteImplAsync(data);
+        }
+        else
+        {
+            /*
+             * In case the number of chunks is bigger than responseMaxChunks we need to make multiple calls 
+             * to the native api https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method  
+             * Despite the documentation states that feeding the function with more than 65535 chunks will cause the function to throw an exception, 
+             * it actually seems that 65534 is the maximum number of chunks allowed.
+             * Also, there seems to be a problem when slicing a ReadOnlySequence on segment borders tracked here https://github.com/dotnet/runtime/issues/67607
+             * That's why responseMaxChunks is set to be 65533. 
+             */
+
+            ushort chunksCount = 0;
+            var start = 0;
+            var length = 0;
+
+            var result = 0;
+
+            foreach (var chunk in data)
+            {
+                chunksCount++;
+                length += chunk.Length;
+
+                if (chunksCount == responseMaxChunks)
+                {
+                    result += await WriteImplAsync(data.Slice(start, length));
+
+                    chunksCount = 0;
+                    start = length;
+                    length = 0;
+                }
+            }
+
+            if (chunksCount > 0)
+            {
+                result += await WriteImplAsync(data.Slice(start, length));
```

Comment:
We cannot just sum HRESULTs like this. [Here's the native implementation.](https://github.com/dotnet/aspnetcore/blob/c7c3fc7d516bc50863b2ba395372ef3537d523e7/src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/managedexports.cpp#L324-L332) We're not currently using the return value of this method anyway, but let's not add these together. Just return the last one or something.

Reply:
Agree, my bad.